pkg/controller/backup/backup_controller.go

```diff
@@ -0,0 +1,119 @@
+	"time"
+
+	hivemetrics "github.com/openshift/hive/pkg/controller/metrics"
+	controllerutils "github.com/openshift/hive/pkg/controller/utils"
+	log "github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/runtime"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+const (
+	controllerName = "backup"
+)
+
+// Add creates a new Backup Controller and adds it to the Manager with default RBAC. The Manager will set fields on the
+// Controller and Start it when the Manager is Started.
+func Add(mgr manager.Manager) error {
+	return AddToManager(mgr, NewReconciler(mgr))
+}
+
+// NewReconciler returns a new reconcile.Reconciler
+func NewReconciler(mgr manager.Manager) reconcile.Reconciler {
+	return &ReconcileBackup{
+		Client: controllerutils.NewClientWithMetricsOrDie(mgr, controllerName),
+		scheme: mgr.GetScheme(),
+		logger: log.WithField("controller", controllerName),
+	}
+}
+
+// AddToManager adds a new Controller to mgr with r as the reconcile.Reconciler
+func AddToManager(mgr manager.Manager, r reconcile.Reconciler) error {
+	// Create a new controller
+	c, err := controller.New(controllerName+"-controller", mgr, controller.Options{Reconciler: r, MaxConcurrentReconciles: controllerutils.GetConcurrentReconciles()})
+	if err != nil {
+		return err
+	}
+
+	reconciler := r.(*ReconcileBackup)
+
+	clusterDeploymentBackup := newClusterDeploymentBackup(reconciler)
+	err = clusterDeploymentBackup.registerClusterDeploymentWatch(c)
+	if err != nil {
+		return err
+	}
+
+	return nil // If we got here, then no errors occurred.
+}
+
+// This ensures that ReconcileBackup struct implements all functions that the reconcile.Reconciler interface requires.
+var _ reconcile.Reconciler = &ReconcileBackup{}
+
+// ReconcileBackup ensures that Velero backup objects are created when changes are made to Hive objects.
+type ReconcileBackup struct {
+	client.Client
+	scheme *runtime.Scheme
+
+	logger log.FieldLogger
+}
+
+// Reconcile ensures that all Hive object changes have corresponding Velero backup objects.
+// ClusterDeployment.Spec.Config.Machines
+// +kubebuilder:rbac:groups=hive.openshift.io,resources=clusterdeployments,verbs=get;list;watch
+// +kubebuilder:rbac:groups=velero.io,resources=backups,verbs=create
+func (r *ReconcileBackup) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	start := time.Now()
+	contextLogger := addReconcileRequestLoggerFields(r.logger, request)
+
+	// For logging, we need to see when the reconciliation loop starts and ends.
+	contextLogger.Info("reconciling backups and Hive object changes")
+	defer func() {
+		dur := time.Since(start)
+		hivemetrics.MetricControllerReconcileTime.WithLabelValues(controllerName).Observe(dur.Seconds())
+		contextLogger.WithField("elapsed", dur).Info("reconcile complete")
+	}()
+
+	clusterDeploymentBackup := newClusterDeploymentBackup(r)
```

Comment:
@twiest Is there any specific reason function `newClusterDeploymentBackup` is not part of `ReconcileBackup` struct i.e. `r. newClusterDeploymentBackup()`